### ISSUE TITLE
[Chrome 51] HTML.TransformingVisitor does not support foreign objects in attributes

### ISSUE THREAD

**Issue by speciose** [meteor/meteor#6793]
I'm using Chrome Canary 51 and this error appear from time to time (pretty rare actualy) when I change route (navigate through my website). I've not seen this error when I was using Meteor 1.2
I don't have any errors on server side.

So. Right now I have Meteor 1.3.1 and Iron Router (latest version with manual update of middleware-stack)

```
iron:controller@1.0.12
iron:core@1.0.11
iron:dynamic-template@1.0.12
iron:layout@1.0.12
iron:location@1.0.11
iron:middleware-stack@1.1.0
iron:router@1.0.12
iron:url@1.0.11
```

And here is the text that I get in chrome console:

```
Exception from Tracker recompute function:
69e39e0….js?meteor_js_resource=true:3 Error: The basic HTML.TransformingVisitor does not support foreign objects in attributes.  Define a custom visitAttributes for this case.
    at visitAttributes (69e39e0….js?meteor_js_resource=true:86)
    at visitAttributes (69e39e0….js?meteor_js_resource=true:90)
    at Object.f._expandAttributes (69e39e0….js?meteor_js_resource=true:90)
    at p (69e39e0….js?meteor_js_resource=true:90)
    at 69e39e0….js?meteor_js_resource=true:90
    at Function.e._withTemplateInstanceFunc (69e39e0….js?meteor_js_resource=true:92)
    at 69e39e0….js?meteor_js_resource=true:90
    at Object.f._withCurrentView (69e39e0….js?meteor_js_resource=true:90)
    at c (69e39e0….js?meteor_js_resource=true:90)
    at o.Computation._compute (69e39e0….js?meteor_js_resource=true:32)
69e39e0….js?meteor_js_resource=true:3 Exception from Tracker recompute function:
69e39e0….js?meteor_js_resource=true:3 TypeError: Cannot read property 'splice' of null
    at https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:90:29128
    at Object.o.nonreactive (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:32:4311)
    at Object.removedAt (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:90:28856)
    at Object.removed (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:88:1831)
    at https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:47:1323
    at Array.forEach (native)
    at Function.A.each.A.forEach (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:1:865)
    at r.diffQueryOrderedChanges (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:47:1281)
    at l (https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:88:1417)
    at https://mywebsite.ru:4488/69e39e041c6e95f5531874ffd10e7adcb29e3623.js?meteor_js_resource=true:88:802
69e39e0….js?meteor_js_resource=true:3 Exception from Tracker recompute function:
69e39e0….js?meteor_js_resource=true:3 TypeError: Cannot read property '0' of null
    at 69e39e0….js?meteor_js_resource=true:90
    at Object.o.nonreactive (69e39e0….js?meteor_js_resource=true:32)
    at Object.changedAt (69e39e0….js?meteor_js_resource=true:90)
    at 69e39e0….js?meteor_js_resource=true:88
    at Function.A.each.A.forEach (69e39e0….js?meteor_js_resource=true:1)
    at l (69e39e0….js?meteor_js_resource=true:88)
    at 69e39e0….js?meteor_js_resource=true:88
    at Object.o.nonreactive (69e39e0….js?meteor_js_resource=true:32)
    at 69e39e0….js?meteor_js_resource=true:88
    at o.Computation._compute (69e39e0….js?meteor_js_resource=true:32)
```

Does it gives you any clues about what goes wrong? After this error I can't change a route and this error produced again and again after I click any link on a page until I refresh (F5) it.

Also. This is production code
Ubuntu 14 LTS 64Bit
Nginx proxy
PM2 project manager
Node v0.10.43

**Comment by speciose** [comment]
Today I get:
```
Exception from Tracker recompute function:
Error: The basic HTML.TransformingVisitor does not support foreign objects in attributes.  Define a custom visitAttributes for this case.
    at visitAttributes (96b8606….js?meteor_js_resource=true:86)
    at visitAttributes (96b8606….js?meteor_js_resource=true:90)
    at Object.f._expandAttributes (96b8606….js?meteor_js_resource=true:90)
    at p (96b8606….js?meteor_js_resource=true:90)
    at 96b8606….js?meteor_js_resource=true:90
    at Function.e._withTemplateInstanceFunc (96b8606….js?meteor_js_resource=true:92)
    at 96b8606….js?meteor_js_resource=true:90
    at Object.f._withCurrentView (96b8606….js?meteor_js_resource=true:90)
    at c (96b8606….js?meteor_js_resource=true:90)
    at o.Computation._compute (96b8606….js?meteor_js_resource=true:32)
```

and hear is that line of code:

```
!function(){var t=Package.underscore._,e,n,r;(function(){e=this}).call(this),function(){n=__meteor_runtime_config__.meteorEnv,r={isProduction:"production"===n.NODE_ENV,isDevelopment:"production"!==n.NODE_ENV,isClient:!0,isServer:!1,isCordova:!1},"object"==typeof __meteor_runtime_config__&&__meteor_runtime_config__.PUBLIC_SETTINGS&&(r.settings={"public":__meteor_runtime_config__.PUBLIC_SETTINGS})}.call(this),function(){function e(t){return t?r._debug("Exception in callback of async function",t.stack?t.stack:t):void 0}if(r.isServer)var n=Npm.require("fibers/future");"object"==typeof __meteor_runtime_config__&&__meteor_runtime_config__.meteorRelease&&(r.release=__meteor_runtime_config__.meteorRelease),t.extend(r,{_get:function(t){for(var e=1;e<arguments.length;e++){if(!(arguments[e]in t))return void 0;t=t[arguments[e]]}return t},_ensure:function(t){for(var e=1;e<arguments.length;e++){var n=arguments[e];n in t||(t[n]={}),t=t[n]}return t},_delete:function(t){for(var e=[t],n=!0,r=1;r<arguments.length-1;r++){var o=arguments[r];if(!(o in t)){n=!1;break}if(t=t[o],"object"!=typeof t)break;e.push(t)}for(var r=e.length-1;r>=0;r--){var o=arguments[r+1];if(n)n=!1;else for(var i in e[r][o])return;delete e[r][o]}},wrapAsync:function(o,i){return function(){for(var a=i||this,s=t.toArray(arguments),u,c=s.length-1;c>=0;--c){var l=s[c],f=typeof l;if("undefined"!==f){"function"===f&&(u=l);break}}if(!u){if(r.isClient)u=e;else{var p=new n;u=p.resolver()}++c}s[c]=r.bindEnvironment(u);var _=o.apply(a,s);return p?p.wait():_}},_inherits:function(e,n){for(var r in n)t.has(n,r)&&(e[r]=n[r]);var o=function(){this.constructor=e};return o.prototype=n.prototype,e.prototype=new o,e.__super__=n.prototype,e}});var o=!1;r._wrapAsync=function(t,e){return o||(r._debug("Meteor._wrapAsync has been renamed to Meteor.wrapAsync"),o=!0),r.wrapAsync.apply(r,arguments)}}.call(this),function(){"use strict";function t(){if(o.setImmediate){var t=function(t){o.setImmediate(t)};return t.implementation="setImmediate",t}return null}function e(){function t(t,e){return"string"==typeof t&&t.substring(0,e.length)===e}function e(e){if(e.source===o&&t(e.data,s)){var n=e.data.substring(s.length);try{a[n]&&a[n]()}finally{delete a[n]}}}if(!o.postMessage||o.importScripts)return null;var n=!0,r=o.onmessage;if(o.onmessage=function(){n=!1},o.postMessage("","*"),o.onmessage=r,!n)return null;var i=0,a={},s="Meteor._setImmediate."+Math.random()+".";o.addEventListener?o.addEventListener("message",e,!1):o.attachEvent("onmessage",e);var u=function(t){++i,a[i]=t,o.postMessage(s+i,"*")};return u.implementation="postMessage",u}function n(){var t=function(t){o.setTimeout(t,0)};return t.implementation="setTimeout",t}var o=this;r._setImmediate=t()||e()||n()}.call(this),function(){var e=function(t){if(Package.ddp){var e=Package.ddp.DDP._CurrentInvocation;if(e.get()&&e.get().isSimulation)throw new Error("Can't set timers inside simulations");return function(){e.withValue(null,t)}}return t},n=function(t,n){return r.bindEnvironment(e(n),t)};t.extend(r,{setTimeout:function(t,e){return setTimeout(n("setTimeout callback",t),e)},setInterval:function(t,e){return setInterval(n("setInterval callback",t),e)},clearInterval:function(t){return clearInterval(t)},clearTimeout:function(t){return clearTimeout(t)},defer:function(t){r._setImmediate(n("defer callback",t))}})}.call(this),function(){r.makeErrorType=function(t,e){var n=function(){Error.captureStackTrace?Error.captureStackTrace(this,n):this.stack=(new Error).stack,e.apply(this,arguments),this.errorType=t};return r._inherits(n,Error),n},r.Error=r.makeErrorType("Meteor.Error",function(t,e,n){var r=this;r.error=t,r.reason=e,r.details=n,r.reason?r.message=r.reason+" ["+r.error+"]":r.message="["+r.error+"]"}),r.Error.prototype.clone=function(){var t=this;return new r.Error(t.error,t.reason,t.details)}}.call(this),function(){r._noYieldsAllowed=function(t){return t()},r._SynchronousQueue=function(){var t=this;t._tasks=[],t._running=!1,t._runTimeout=null},t.extend(r._SynchronousQueue.prototype,{runTask:function(e){var n=this;if(!n.safeToRunTask())throw new Error("Could not synchronously run a task from a running task");n._tasks.push(e);var o=n._tasks;n._tasks=[],n._running=!0,n._runTimeout&&(clearTimeout(n._runTimeout),n._runTimeout=null);try{for(;!t.isEmpty(o);){var i=o.shift();try{i()}catch(a){if(t.isEmpty(o))throw a;r._debug("Exception in queued task: "+(a.stack||a))}}}finally{n._running=!1}},queueTask:function(e){var n=this;n._tasks.push(e),n._runTimeout||(n._runTimeout=setTimeout(t.bind(n.flush,n),0))},flush:function(){var t=this;t.runTask(function(){})},drain:function(){var e=this;if(e.safeToRunTask())for(;!t.isEmpty(e._tasks);)e.flush()},safeToRunTask:function(){var t=this;return!t._running}})}.call(this),function(){var t=[],e=!1,n=!1,o=0,i=function(){o++},a=function(){o--,s()},s=function(){if(!(n||!e||o>0)){for(n=!0;t.length;)t.shift()();r.isCordova&&WebAppLocalServer.startupDidComplete()}},u=function(){e||(e=!0,s())};r.isCordova&&(i(),document.addEventListener("deviceready",a,!1)),"complete"===document.readyState||"loaded"===document.readyState?window.setTimeout(u):document.addEventListener?(document.addEventListener("DOMContentLoaded",u,!1),window.addEventListener("load",u,!1)):(document.attachEvent("onreadystatechange",function(){"complete"===document.readyState&&u()}),window.attachEvent("load",u)),r.startup=function(e){var o=!document.addEventListener&&document.documentElement.doScroll;if(o&&window===top){try{o("left")}catch(i){return void setTimeout(function(){r.startup(e)},50)}e()}else n?e():t.push(e)}}.call(this),function(){var t=0;r._debug=function(){if(t)return void t--;if("undefined"!=typeof console&&"undefined"!=typeof console.log)if(0==arguments.length)console.log("");else if("function"==typeof console.log.apply){for(var e=!0,n=0;n<arguments.length;n++)"string"!=typeof arguments[n]&&(e=!1);e?console.log.apply(console,[Array.prototype.join.call(arguments," ")]):console.log.apply(console,arguments)}else if("function"==typeof Function.prototype.bind){var r=Function.prototype.bind.call(console.log,console);r.apply(console,arguments)}else Function.prototype.call.call(console.log,console,Array.prototype.slice.call(arguments))},r._suppress_log=function(e){t+=e},r._suppressed_log_expected=function(){return 0!==t}}.call(this),function(){r._escapeRegExp=function(t){return String(t).replace(/[.*+?^${}()|[\]\\]/g,"\\$&")}}.call(this),function(){var t;t=r.isClient?n.TEST_METADATA:process.env.TEST_METADATA;var e=JSON.parse(t||"{}"),o=e.driverPackage;r.isTest=!!e.isTest,r.isAppTest=!!e.isAppTest,r.isPackageTest=!!o&&!r.isTest&&!r.isAppTest,"string"==typeof o&&r.startup(function(){var t=Package[o];if(!t)throw new Error("Can't find test driver package: "+o);if(r.isClient){if("function"!=typeof t.runTests)throw new Error("Test driver package "+o+" missing `runTests` export");t.runTests()}else"function"==typeof t.start&&t.start()})}.call(this),function(){var e=0,n=[];r.EnvironmentVariable=function(){this.slot=e++},t.extend(r.EnvironmentVariable.prototype,{get:function(){return n[this.slot]},getOrNullIfOutsideFiber:function(){return this.get()},withValue:function(t,e){var r=n[this.slot];try{n[this.slot]=t;var o=e()}finally{n[this.slot]=r}return o}}),r.bindEnvironment=function(e,o,i){var a=t.clone(n);if(!o||"string"==typeof o){var s=o||"callback of async function";o=function(t){r._debug("Exception in "+s+":",t&&t.stack||t)}}return function(){var r=n;try{n=a;var s=e.apply(i,t.toArray(arguments))}catch(u){o(u)}finally{n=r}return s}},r._nodeCodeMustBeInFiber=function(){}}.call(this),function(){r.absoluteUrl=function(e,n){n||"object"!=typeof e||(n=e,e=void 0),n=t.extend({},r.absoluteUrl.defaultOptions,n||{});var o=n.rootUrl;if(!o)throw new Error("Must pass options.rootUrl or set ROOT_URL in the server environment");return/^http[s]?:\/\//i.test(o)||(o="http://"+o),/\/$/.test(o)||(o+="/"),e&&(o+=e),n.secure&&/^http:/.test(o)&&!/http:\/\/localhost[:\/]/.test(o)&&!/http:\/\/127\.0\.0\.1[:\/]/.test(o)&&(o=o.replace(/^http:/,"https:")),n.replaceLocalhost&&(o=o.replace(/^http:\/\/localhost([:\/].*)/,"http://127.0.0.1$1")),o},r.absoluteUrl.defaultOptions={},"object"==typeof __meteor_runtime_config__&&__meteor_runtime_config__.ROOT_URL&&(r.absoluteUrl.defaultOptions.rootUrl=__meteor_runtime_config__.ROOT_URL),r._relativeToSiteRootUrl=function(t){return"object"==typeof __meteor_runtime_config__&&"/"===t.substr(0,1)&&(t=(__meteor_runtime_config__.ROOT_URL_PATH_PREFIX||"")+t),t}}.call(this),"undefined"==typeof Package&&(Package={}),function(t,e){for(var n in e)n in t||(t[n]=e[n])}(Package.meteor={},{Meteor:r,global:e,meteorEnv:n})}();
```

**Comment by speciose** [comment]
Can't reproduce this error on development. So for now it only exist in production

**Comment by speciose** [comment]
Could the problem be with underscore?

**Comment by macroramesh6** [comment]
I have same issues after updated to 1.3
````
Exception from Tracker recompute function: undefined
debug.js:43 Error: The basic HTML.TransformingVisitor does not support foreign objects in attributes.  Define a custom visitAttributes for this case.
    at HTML.TransformingVisitor.def.visitAttributes (http://localhost:3000/packages/htmljs.js?hash=3035d42f05f842d4beb2111cc6739b9ce6cd4ca4:209:13)
    at Blaze._HTMLJSExpander.def.visitAttributes (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:2136:63)
    at HTML.TransformingVisitor.def.visitAttributes (http://localhost:3000/packages/htmljs.js?hash=3035d42f05f842d4beb2111cc6739b9ce6cd4ca4:197:44)
    at Blaze._HTMLJSExpander.def.visitAttributes (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:2136:63)
    at Object.Blaze._expandAttributes (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:2165:32)
    at updateAttributes (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:1520:33)
    at http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:1875:20
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:3687:12)
    at http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:1873:29
    at Object.Blaze._withCurrentView (http://localhost:3000/packages/blaze.js?hash=ef41aed769a8945fc99ac4954e8c9ec157a88cea:2214:12) undefined
````

**Comment by laosb** [comment]
Would you like to try run development with `--production`?

**Comment by macroramesh6** [comment]
@laosb Yes it produces same error with `--production`

**Comment by laosb** [comment]
Sounds like a bug of iron-router.

**Comment by macroramesh6** [comment]
@laosb I am using `Flow router`

**Comment by laosb** [comment]
So that means a Meteor bug? Let me see...

**Comment by brianlukoff** [comment]
This is a serious problem -- I'm seeing the same thing, in production, with Meteor 1.3.2.4, Flow Router, and Chrome 51 (which is now being pushed out to everyone).

**Comment by abernix** [comment]
If you're experiencing this problem, please create a basic reproduction repo which exhibits the problem and post a link to it in this issue so it can be easily cloned and the error exhibited.

@brianlukoff said:
> I'm seeing the same thing, in production, with Meteor 1.3.2.4, Flow Router, and Chrome 51 (which is now being pushed out to everyone).

I just navigated routes a lot on a good size project with these same exact conditions (though you didn't specify your FlowRouter version, I'm using `2.12.1`) and I'm **not** seeing this problem. None of my Chimp acceptance tests running in Chrome seem to be showing it either.

**Comment by abernix** [comment]
Also, check your code to make sure you're not doing something that Blaze doesn't support.  This actual error is [coming from these lines](https://github.com/meteor/meteor/blame/1f61491f458436954e845689912a0d9a83755d46/packages/htmljs/visitors.js#L135-L160).

Make sure you're not passing a `new Foo` object instead of just `{ bar: true }` as attributes on HTML tags.  [This htmljs README page gives some further explanation](https://github.com/meteor/meteor/blob/1f61491f458436954e845689912a0d9a83755d46/packages/htmljs/README.md#htmlisconstructedobjectx).

**Comment by brianlukoff** [comment]
Thanks for taking a look at this! It's very strange -- I haven't even been able to reproduce in development (even when I start meteor with the --production flag!), it only happens on Chrome, and it only happens on a single route. (I'm using FlowRouter 2.12.1 as well.)  It doesn't happen on every request; probably about 1/3-1/2 of the time.

Debugging using Chrome Developer Tools in production, it looks like `attrs` on [this line](https://github.com/meteor/meteor/blame/1f61491f458436954e845689912a0d9a83755d46/packages/htmljs/visitors.js#L156) is set to an object that itself includes both `attrs` and `children` properties, while on development this is not the case (if in production `attrs` is `{ attrs: X, children: Y }`, then on development `attrs` just seems to be `X`).

**Comment by brianlukoff** [comment]
I've tracked this down to a single `<a>` tag in a Bootstrap dropdown in a header template. The only thing special about that link is that it points to an external site (rather than to an internal route in the app). Removing that tag resolves the issue, but it's not clear why!

**Comment by speciose** [comment]
I'm back. Are you using 32bit or 64bit Chrome? Because I had this issue only on 64bit. After that I've downgraded to stable 32bit Chrome (ver 50) and everything become great.
I've updated today to ver 51. I'll let you know if this error will happen again

**Comment by brianlukoff** [comment]
I'm on 64-bit.

**Comment by toszter** [comment]
My hunch is that `pathFor` is somehow responsible. I'm using it in anchors throughout tabular data.  The darndest thing is that I'm supposed to present the live search front-end we wrote using Meteor to the CEO of our company tomorrow... and now this! <sighs>

**Comment by abernix** [comment]
@toszter Information about your environment would be appreciated – in addition to details about when you first saw this error (were other packages updated when you upgraded?) and how you've narrowed it down to `pathFor` (which is a FlowRouter helper).

And of course, if someone can narrow this down to a simple reproduction, that would be most beneficial for getting it resolved.

**Comment by toszter** [comment]
@abernix I did no upgrades to Meteor.  Google Chrome just auto-updated.  I'm only suspecting `pathFor` because the client view that is crashing uses it exclusively for all anchor tags.

Honestly, my primary concern is with the product that I've built using Meteor.  When I get that working, I'll help out w/ Meteor. In the meantime, if anyone else is reading this with interest, then please take these steps to repro (or at least test):

1. Download and install Meteor 1.3.2.4
2. Create a route for any collection item of your choice
3. Create a client template with Blaze that iterates the client collection items
4. Use `pathFor` in the template iterator that points to the item route

I'm currently removing all the `pathFor` declarations in my client side templates.

**Comment by toszter** [comment]
**Update:** - I removed all the occurrences of `pathFor` as well as any external links from the `href` in anchors throughout my codebase, as @brianlukoff had mentioned was occurring in his example. The error is still there.

Now, I'm going to start rolling the Meteor version back to 1.2.1 and re-deploying each into Galaxy until I see the error again... my hope is that I'll discover the breaking revision and we'll see the diff...

**Comment by brianlukoff** [comment]
Thanks @toszter - when I removed the `<a>` tag that I thought was causing the problem, the error disappeared for me (and I still can't get it to recur in manual testing in production) but I can see in our error tracking that it is still occurring for some users. So I'm stuck again. Looking forward to the results of your analysis!

**Comment by toszter** [comment]
**Update:** I'm unable to rollback to 1.2.1 from 1.3.2.4 so I rolled back to 1.3.1...which _still_ has the error.  So, now I'm unsure what to do, but I imagine looking at Meteor sources and polishing my résumé are a good start.

**Comment by brianlukoff** [comment]
It could be some long-standing code in Meteor that is just happening to interact with a change in Chrome 51? Have you been able to reduce this to a smaller reproduction repo? I haven't been able to reproduce the problem in a smaller repo yet.

**Comment by toszter** [comment]
I haven't attempted that. Is there a way to deploy meteor apps to production in Galaxy w/o incurring a charge? I only have access to my production Galaxy instances. I suppose I could overwrite my production code with a small repro instance, but that's so not ideal...  @abernix is there a test account we could use at MDG for that?

Also, what additional packages are you using (the output of `meteor list`)? Mine is:

```
accounts-password           1.1.8  Password support for accounts
accounts-ui                 1.1.9  Simple templates to add login widgets to an app
alanning:roles              1.2.15  Authorization package for Meteor
aldeed:autoform             5.8.1  Easily create forms with automatic insert and update, and automatic reactive validation.
aldeed:collection2          2.9.1  Automatic validation of insert and update operations on the client and server.
aldeed:simple-schema        1.5.3  A simple schema validation object with reactivity. Used by collection2 and autoform.
arillo:flow-router-helpers  0.5.1  Template helpers for flow-router
blaze-html-templates        1.0.4  Compile HTML templates into reactive UI with Meteor Blaze
coffeescript                1.0.17  Javascript dialect with fewer braces and semicolons
ecmascript                  0.4.3  Compiler plugin that supports ES2015+ in all .js files
ejson                       1.0.11  Extended and Extensible JSON library
email                       1.0.12  Send email messages
es5-shim                    4.5.10  Shims and polyfills to improve ECMAScript 5 support
force-ssl                   1.0.11  Require this application to use HTTPS
fourseven:scss              3.4.3  Style with attitude. Sass and SCSS support for Meteor.js (with autoprefixer and sourcemaps).
http                        1.1.5  Make HTTP calls to remote servers
jquery                      1.11.8  Manipulate the DOM using CSS selectors
kadira:blaze-layout         2.3.0  Layout Manager for Blaze (works well with FlowRouter)
kadira:flow-router          2.12.1  Carefully Designed Client Side Router for Meteor
meteor-base                 1.0.4  Packages that every Meteor app needs
meteorhacks:async           1.0.0  Set of async utilities to work with NPM modules and other async code blocks
meteorhacks:npm             1.5.0  Use npm modules with your Meteor App
mobile-experience           1.0.4  Packages for a great mobile user experience
momentjs:moment             2.13.1  Moment.js (official): parse, validate, manipulate, and display dates - official Meteor packaging
mongo                       1.1.7  Adaptor for using MongoDB and Minimongo over DDP
mrgalaxy:stripe             2.2.2  Stripe.js and Node-Stripe brought to Meteor.
npm-container               1.2.0+ Contains all your npm dependencies
peppelg:bootstrap-3-modal   1.0.4  Simple usage of bootstrap 3 modals.
random                      1.0.9  Random number generator and utilities
reactive-var                1.0.9  Reactive variable
service-configuration       1.0.9  Manage the configuration for third-party services
session                     1.1.5  Session variable
standard-minifier-css       1.0.6  Standard css minifier used with Meteor apps by default.
standard-minifier-js        1.0.6  Standard javascript minifiers used with Meteor apps by default.
tracker                     1.0.13  Dependency tracker to allow reactive callbacks
twbs:bootstrap              3.3.6  The most popular front-end framework for developing responsive, mobile first projects on the web.
```

**Comment by brianlukoff** [comment]
I'm using all of those packages except accounts-ui, coffeescript, fourseven:scss, meteorhacks:async, meteorhacks:npm, mobile-experience, mrgalaxy:stripe, peppelg:bootstrap-3-modal, and twbs:bootstrap.

I don't think Meteor has a free hosting option anymore, but you can use [mupx](https://github.com/arunoda/meteor-up/tree/mupx) to deploy to a $5 DigitalOcean droplet.

**Comment by toszter** [comment]
thanks @brianlukoff, _cough cough_ MDG needs to pony up an ephemeral dev host for community contributions... 

I'm now looking at the upstream invoking code wondering how to go about tracing the problem with an un-minimized meteor source...in true production on Galaxy? Any suggestions?

The method in question is invoked on one of these locations:
```
./packages/blaze/view.js:454:  visitAttributes: function (attrs) {
./packages/blaze/view.js:460:    return HTML.TransformingVisitor.prototype.visitAttributes.call(this, attrs);
./packages/blaze/view.js:489:    {parentView: parentView})).visitAttributes(attrs);
./packages/htmljs/visitors.js:123:    var newAttrs = this.visitAttributes.apply(this, argsCopy);
./packages/htmljs/visitors.js:138:  visitAttributes: function (attrs/*, ...*/) {
./packages/htmljs/visitors.js:145:        var newItem = this.visitAttributes.apply(this, argsCopy);
./packages/htmljs/visitors.js:159:                      "visitAttributes for this case.");
./packages/spacebars-compiler/compiler.js:35:  visitAttributes: function (attrs) {
./packages/spacebars-compiler/compiler.js:40:    return HTML.TransformingVisitor.prototype.visitAttributes.call(this, attrs);
./packages/spacebars-compiler/optimizer.js:66:    if (this.visitAttributes(tag.attrs) !== OPTIMIZABLE.FULL)
./packages/spacebars-compiler/optimizer.js:71:  visitAttributes: function (attrs) {
./packages/spacebars-compiler/optimizer.js:97:  visitAttributes: function (attrs/*, ...*/) {
./packages/spacebars-compiler/optimizer.js:102:    return HTML.TransformingVisitor.prototype.visitAttributes.apply(
./packages/spacebars-compiler/optimizer.js:139:  visitAttributes: function (attrs) {
```

**Comment by abernix** [comment]
I don't think this is a Galaxy problem.  But why not just run it for free, "in production" on your local machine?  [Instructions are here](http://guide.meteor.com/deployment.html#custom-deployment) and there's nothing different about the node version you can run on your local machine versus Galaxy.  Just make sure you're using the right version of Node (0.10.43 if you're on 1.3.2.4).

Unfortunately, the reproduction steps I've seen describe my live projects as well – and I'm not seeing this error.   To me, it doesn't appear to be the `pathFor` method (which is provided by flow-router-helpers), it appears to be an object that's being passed as an argument to an HTML tag in a Blaze template. (e.g. `<i {{someHelperReturningSomethingStrange}}></i>`).  As if Blaze was trying to attribute-ize something that is not an vanilla key/value object (like a collection, a class, a null value, etc.).

I'm most curious if anyone can reproduce this in FireFox, Safari or Chrome 50.

**Comment by abernix** [comment]
And in addition, I'd like to know if it happens in the *New* Canary, v53.  If it's gone in 53, then we could perhaps focus on this being a temporary Chrome bug, and try to find a work-around.

**Comment by derouck** [comment]
I had a broken application with chrome 51 as well. After some tinkering I found the main reason was in the iron-router package. This upgrade was made specifically to solve this problem I think: iron:middleware-stack@1.1.0 

It looks like it's not the same issue everyone has here as most of you are using Flow Router, but this old project for me was running Meteor 1.2.0 and Iron router.

Allthough it might give you a hint to where to look as maybe the same root problem is causing the issue. You'll see the last commit made for this middleware stack is the only one since long to solve a problem with 51.

Good luck!

**Comment by Nexus7** [comment]
@derouck Also having the same problem as you... Production Meteor application which uses Iron Router which started to break with the new Chrome 51.x.x updates... Unfortunately we are not currently in a position whereby it is easy for us to rebuld and redploy our application (due to current development work). Is there any quick hack we can do in production to fix this without a full rebuild/redploy cycle?

Cheers!

**Comment by derouck** [comment]
@Nexus7 I bet there are gonna be hundreds of applications that will be affected. The minimum change I needed to do was quite simple:

`meteor update iron:middleware-stack`

If that command is updating the package to version 1.1.0 you might be good after a simple deploy.

**Comment by Nexus7** [comment]
Thanks @derouck. But I think we can only do a Meteor update on that package on an app which has not yet been built? In which case we would still need to run a build/deploy step? The issue there is that we currently have another package issue with building at this time... So we are kind of between a rock and a hard place right now 😥

**Comment by toszter** [comment]
@abernix - Thanks for those instructions. However, they're for an npm-initialized Meteor installation. That's not something we've done yet over here though we're on 1.3.2.4 (we're still awaiting a stable node.js v4 release that supports MongoDB v3.2.x w/ WiredTiger) so it'll be boiling the ocean to trace this using that method, unless you can enlighten me further.

Instead, I think a clever trick might be to add a `console.trace()` and a big thick `console.log()`  to the Meteor sources [on line 155](https://github.com/meteor/meteor/blame/1f61491f458436954e845689912a0d9a83755d46/packages/htmljs/visitors.js#L156) and any predicating method calls as well to get a comprehensive breakdown of attributes in the chain.  

Only then can we compile and deploy the traced code into Galaxy to get an accurate snapshot rather than add `--production` to a development environment to "mimic" production behavior at any rate.  

I won't be able to try that until tomorrow at the earliest so if anyone else wants to take a "dare", by all means... 🎱

**Comment by mrw34** [comment]
In case it provides any useful information: we're seeing this issue but with a self-hosted (i.e. non-Galaxy) app, which is running Meteor 1.2 and Iron Router 0.9(!).

**Comment by abernix** [comment]
Ok, it seems there are now at least two, but possibly three issues going on in here.  Hopefully we can get this back on track.  :smile:

If you're encountering the `iron-router` issue it seems you'd be getting an error roughly like this: `Error: Handler with name 'somethingHere' already exists. `  Please make a new issue if you're encountering that.  Unfortunately, it sounds like it will bite some people, but the fix is easy: `meteor update iron:middleware-stack`. @Nexus7 You should branch your Git repo at the currently deployed commit, make a hot-fix with this command, then deploy that.

@toszter First of all, if you're not getting the error that is in the subject of this issue (`Cannot read property 'splice' of null`), I'm going to have you open a new, separate issue for the `HTML.TransformingVisitor` error.  But in general, those are still the same instructions irregardless of whether or not you take advantage of NPM.  You can skip the `npm install --production` before building, but running it won't do anything if you don't have a `package.json`.  You should still be doing an `npm install` on the server inside the `programs/server` directory – that hasn't changed in a long time.  Also I don't think that `console.log`-ing is necessary – @brianlukoff already showed that he got the values with DevTools.  The more helpful method to look at is [in this code](https://github.com/meteor/meteor/blob/38a119f14d45518d7e7a73098dc033585cb862ff/packages/htmljs/html.js#L180-L192) which shows what `HTML.isConstructedObject` is checking for.  Specifically, it checks for if you are passing a `Class` or an object created with `new Object({params: true });` as an attribute to any HTML element.

@mrw34 Please be more specific about what issue you're seeing.  If it's not the exact one in the subject of this issue, please open it in a separate ticket.

**Comment by mrw34** [comment]
More specifically we've been seeing the following since the release of Chrome 51. Occurs when switching between pages of our app but I can't reproduce reliably:
```
Error: The basic HTML.TransformingVisitor does not support foreign objects in attributes.  Define a custom visitAttributes for this case.
    at visitAttributes (1163794….js?meteor_js_resource=true:85)
    at visitAttributes (1163794….js?meteor_js_resource=true:91)
    at Object.d._expandAttributes (1163794….js?meteor_js_resource=true:91)
    at h (1163794….js?meteor_js_resource=true:91)
    at 1163794….js?meteor_js_resource=true:91
    at Function.e._withTemplateInstanceFunc (1163794….js?meteor_js_resource=true:92)
    at 1163794….js?meteor_js_resource=true:91
    at Object.d._withCurrentView (1163794….js?meteor_js_resource=true:91)
    at l (1163794….js?meteor_js_resource=true:91)
    at t.Computation._compute (1163794….js?meteor_js_resource=true:33)
```

**Comment by abernix** [comment]
Ok, since nobody has posted about the `splice` error since the original post (including the original author), I'm hijacking this Issue and making it about the `HTML.TransformingVisitor` issue. :smile:

@toszter Never mind opening in a new issue as I posted in my (now edited) comment.  Please do report about whether you're getting the "`Cannot read property 'splice' of null`" error though – it doesn't sound like you are.

@mrw34 Please try to reliably reproduce if you can.  What router are you using?

**Comment by mrw34** [comment]
Iron Router 0.9 on Meteor 1.2.

**Comment by brianlukoff** [comment]
Is it possible to add the labels severity:production and impact:some to this issue?

**Comment by abernix** [comment]
@brianlukoff Definitely.  I meant to upgrade that when I hijacked it earlier, thanks for the shout.

**Comment by abernix** [comment]
Also, since apparently this is happening on Meteor 1.2, I'm removing the version from the title and adding [Chrome 51].  Still curious if anyone can repro in FireFox, FireFox Dev Edition or Safari.

**Comment by brianlukoff** [comment]
Thanks @abernix!

**Comment by micktaiwan** [comment]
Appearing only in production mode, also with Iron:Router, also with Meteor 1.3.3

**Comment by JonathonJulian** [comment]
Just started seeing the same behavior today, updating iron did not help, no issues in Firefox or Chrome from localhost, problem is only seen in prod hosted on Galaxy.

**Comment by nathan-muir** [comment]
Having this also in 2 apps using Meteor@1.2 on Chrome 51

It seems completely random, and I can't reproduce it repeating the same actions that caused an observed error.

If it's not something in my applications code, I just wonder what could've broken the `isConstructedObject` check in chrome, in some random way.

```js
isConstructedObject = function(t) {
    return t && "object" == typeof t && t.constructor !== Object && "function" == typeof t.constructor && t instanceof t.constructor
}
```

Maybe some weird object re-use/garbage collection bug?

**Comment by hazio** [comment]
Having this also in my app using Meteor@1.3.3 on Chrome 51.0.2704.84 (64-bit).

**Comment by brianlukoff** [comment]
Is it possible for someone at MDG to take a look at this -- it seems like this is fairly widespread (see also [this related forum post](https://forums.meteor.com/t/chrome-51-error-the-basic-html-transformingvisitor-does-not-support-foreign-objects-in-attributes/23961/9)).

**Comment by llama** [comment]
I'm having this problem in my app. I've been telling users to use Firefox.

**Comment by brianlukoff** [comment]
I can confirm it is still an issue in 1.3.3.1.

**Comment by elie222** [comment]
Just encountered it on a production app running Meteor 1.3.2.4 and Iron Router on a Windows 7 machine using Chrome 51.0.2704.84.

A refresh fixed it. It happens randomly.

**Comment by tobias-feldmann** [comment]
Same problem on a own production machine and only with Chrome. Localhost runs fine.
Chrome version 51.0.2704.103 (64-bit)
meteor@1.1.10
iron:router@1.0.12
iron:middleware-stack@1.1.0

**Comment by tmeasday** [comment]
@tobias-feldmann any chance of getting a reproduction? I'm sure people can appreciate it's very difficult to approach this kind of problem without one.

**Comment by tmeasday** [comment]
For those still seeing this issue, can you please post a little more information so we can try and track this down? Obviously a reproduction we can run would be ideal.

1. Are you only seeing the issue on production hosting? Can you make it occur with `meteor --production` or running a production bundle locally produced with `meteor bundle`? (See @abernix's [comment](https://github.com/meteor/meteor/issues/6793#issuecomment-222614432))

2. Do you only see the issue when changing routes? [1]

3. What does `t` look like if you set a breakpoint at the `visitAttributes` function in your production app? 

  [To do this, try expanding the minified source, searching for the term "basic HTML.TransformingVisitor" and setting a breakpoint in the chrome inspector. You should see something like this: https://www.dropbox.com/s/uh1uoq5qxxtlb4h/Screenshot%202016-06-21%2012.26.31.png?dl=0. What we are interested in is the value of `t` right before the exception is thrown, so setting the breakpoint inside the `if` statement would be the best way to get that: https://www.dropbox.com/s/cqcqqyazxn0i00f/Screenshot%202016-06-21%2012.26.58.png?dl=0 [2])

4. What does the snippet of Blaze/helpers that trigger the call referenced above?

   The code that would trigger this would look something like `<div {{attrs}}>`, where `attrs` is a helper. You might be able to reverse engineer which specific case that is from the `t` above. If so can you post the blaze tag and helper code here.

5. Have you tried to reproduce with Chrome Canary?

6. What does you package list look like (can you please just post a link to a gist, let's not make this thread unnecesarily longer.

[1] Given it happens for both IR and FR I'd suggest it's unlikely to be a routing specific issue, probably just that changing routes causes a lot of re-rendering.

[2] I can't reproduce the problem so I have set the breakpoint outside.

**Comment by brianlukoff** [comment]
Thanks Tom -- here's what I'm seeing in my app:

1. I'm only seeing it in production apps (deployed on my own servers, not on Galaxy). I can't reproduce using `--production`. I can't even reproduce consistently, but if I reload enough times and change routes enough I can get it to happen.
2. I only see the issue when changing routes.
3. Here's [a screenshot](https://www.dropbox.com/s/00qj5retp8spdfv/Screenshot%202016-06-20%2022.11.00.png?dl=0) of what `t` looks like.
4. The Blaze that's being triggered is a link: `<a href="{{pathFor 'users'}}">Manage users</a>`.  The `pathFor` helper is from `arillo:meteor-flow-router-helpers`.
5. Yes, I can reproduce in Chrome Canary.

**Comment by nathan-muir** [comment]
@tmeasday 

1. I've seen it happen in two separate apps, once on a workmates PC, and once on my own. Both OSX using chrome 51. 

Here's a list of the affected OS's from stack traces in Kadira-
```
Chrome (WebKit) - 51.0.2704.84  Windows - 7
Chrome (WebKit) - 51.0.2704.103  Windows - 7
Chrome (WebKit) - 51.0.2704.103  Windows - 8.1
Chrome (WebKit) - 51.0.2704.84  Windows - 10
Chrome (WebKit) - 51.0.2704.79  Chromium OS - 8172.47.0
Chrome (WebKit) - 51.0.2704.103  Mac OS - 10.10.5
Chrome (WebKit) - 51.0.2704.84  Mac OS - 10.11.4
```

2. Once routing, another time opening a modal.

3. Setting the breakpoint was my first step, however I've not yet been able to reproduce it, even after fuzzing with my mouse/keyboard for 15-20mins.

4/5/6. Unknown as yet. / Not Yet / Need to redact it before sharing.

**Comment by tmeasday** [comment]
@brianlukoff perhaps you can dig a little deeper if you can reliably get into that break point.

A. What is `t.constructor`?
B. If you step up the stack trace to the function that calls `expandAttributes` (called `p` in the stacktrace at the top), does it look like this: https://www.dropbox.com/s/lw0b805ph50n4ps/Screenshot%202016-06-21%2015.52.46.png?dl=0 ? (ie. is it this function: https://github.com/meteor/meteor/blob/fix-pr-6889/packages/blaze/materializer.js#L92)
C. Are you able to get any references to variables in that function (specifically the first argument, `tag`)?

   (You could try putting a conditional breakpoint in that function if the `rawAttrs` variable has a `children` property)

**Comment by brianlukoff** [comment]
@tmeasday Here's what I found:

A. Inspecting `t.constructor` gives this: 
![image](https://cloud.githubusercontent.com/assets/518013/16233908/858a67ec-3795-11e6-9a96-5454d445004a.png)
which points to this function:
![image](https://cloud.githubusercontent.com/assets/518013/16233965/a7639fc8-3795-11e6-9889-73ed91519ad1.png)

B. Yes, I'm getting what looks like the `materializeTag` function.

C. `tag` and `workStack` are undefined, and `parentView` is:
![image](https://cloud.githubusercontent.com/assets/518013/16234688/be0560ca-3797-11e6-8c16-1863b72b979d.png)

Let me know what other info would be helpful!

**Comment by tobias-feldmann** [comment]
1. I can reproduce it on localhost with meteor --production
2. Currently only after routing to a page with a lot of views which contains this code:

```
<ul class="dropdown-menu" id="typeDropdownMenu">
	{{#each photoTypes }}
		{{> photoTypeItem}}
	{{/each}}
</ul>

<template name="photoTypeItem">
	<li><a href="#">{{text}}</a></li>
</template>
```

photoTypes is a array composed of objects ({'text': pricesText};)
If if use fixed text in place of {{text}} then it works. 

The problem occurs almost every time if i have plenty of this views.
With only a few i get it really seldom. 



This is the generated code for this view:
```
Template.__checkName("photoTypeItem"),Template.photoTypeItem=new Template("Template.photoTypeItem",function(){var n=this;return HTML.LI(HTML.A({href:"#"},Blaze.View("lookup:text",function(){return Spacebars.mustache(n.lookup("text"))})))})
```

3. And the debug stuff:

https://www.dropbox.com/s/qv2dx8la7d33o7p/debug1.png?dl=0

https://www.dropbox.com/s/j3kx9ak1c2oqxxm/debug2.png?dl=0

**Comment by mwoodbri** [comment]
Exactly same for us - the only problem in our app is for this dropdown, where `allOrganisations` is a long list of objects:

    <ul class="dropdown-menu">
      {{#each allOrganisations}}
      <li><a href="#">{{Name}}</a></li>
      {{/each}}
    </ul>

**Comment by ecwyne** [comment]
Same for me here

```html
<ul class="dropdown-menu" role="menu">
	{{#each users}}
	    <li><a href="#" class="assignToBtn" data-id={{_id}}>{{profile.firstName}}</a></li>
	{{/each}}
</ul>
```

**Comment by brianlukoff** [comment]
Interesting -- the link I've found that triggers the issue is also in a Boostrap `dropdown-menu`!

**Comment by tobias-feldmann** [comment]
I debugged the materializeTag function in the materializer.js with both states.

Here is the state which crashed after that.
The materializeTag function get a wrong tag (e) which contains the attrs directly.
https://www.dropbox.com/s/de95utb8h2mmqis/debug3.png?dl=0
https://www.dropbox.com/s/a9o8u6anug7b30k/debug5.png?dl=0

Here a correct state where we have the attrs in the children array.
https://www.dropbox.com/s/y9vpiq8wphzxwdx/debug4.png?dl=0
https://www.dropbox.com/s/ssbppx83xi7tr67/debug6.png?dl=0

**Comment by ecwyne** [comment]
@tmeasday 

Here's the smallest reproduction I could find

https://github.com/ecwyne/TransformingVisitor-Bug

**Comment by tmeasday** [comment]
Great work on the reproduction! Thankyou @ecwyne.

OK, this is a bizarre one. I see the issue in Chrome 51 **but not in Chrome 53** so I suspect it is just at root a chrome bug. @brianlukoff you said you still see it in your app with Chrome Canary? I wonder which version of Chrome that is?

I traced through this and ultimately the issue seems to be in this code: https://github.com/meteor/meteor/blob/fix-pr-6889/packages/htmljs/html.js#L12-L44

Sometimes, when running code as simple as `HTML.LI(HTML.A())`, it enters the branch where it thinks the `A()` is an attrs object and so you end up with something that looks like `{ attrs: { attrs: {...} }`, which is never correct.

However, even a simple reproduction like the above is unreliable, and chrome is acting really weird. I was running:

```
_.filter(_.times(1000, () => HTML.LI(HTML.A())), li => li.children.length > 0)
```

And seeing results as varied as `[]`, `Array[1000]`, and even sometimes things like `Array[743]`.

But I *do* have a work around. Simply adding some whitespace inside the `<li>` so that it has more than one child is enough to mean that @ecwyne's reproduction no longer breaks ever for me.

Let me know if people are seeing this on Chrome > 51, and if the workaround doesn't help.

**Comment by ecwyne** [comment]
Thanks @tmeasday !
When I first ran some tests I also see that adding some whitespace BEFORE the `<a>` tag removes the error for me, but it does not work if I put whitespace AFTER the `<a>` tag

```html
<!-- works -->
<ul class="dropdown-menu">
  {{#each items}}
    <li>@nbsp;<a href="/{{this}}">{{this}}</a></li>
  {{/each}}
</ul>

<!-- does not work -->
<ul class="dropdown-menu">
  {{#each items}}
    <li><a href="/{{this}}">{{this}}</a>@nbsp;</li>
  {{/each}}
</ul>
```

**Comment by tmeasday** [comment]
That actually makes sense @ecwyne when you consider the implementation of the `HTMLTag` function I referenced above.

**Comment by nathan-muir** [comment]
https://www.reddit.com/r/javascript/comments/4oxtgk/javascript_developers_be_warned_about_this_crazy/

**Comment by mwoodbri** [comment]
Great to hear about the Chromium patch. In the meantime the workaround has solved the issue for us too. Thanks everyone, especially @tmeasday!

**Comment by brianlukoff** [comment]
Thanks @tmeasday!  The Canary version I was using was Version 53.0.2774.3.

**Comment by lachesis** [comment]
I still see this in our production app in Chrome Version 53.0.2774.3 dev (64-bit) on Arch Linux. However, I do not see it in the minimal repro case.

Is this a precise statement of the workaround? "Any tag with a dynamic attribute must have at least one sibling."

**Comment by tmeasday** [comment]
> The Canary version I was using was Version 53.0.2774.3

> I still see this in our production app in Chrome Version 53.0.2774.3 dev (64-bit) on Arch Linux. However, I do not see it in the minimal repro case.

Hmm, OK, it sounds like Chrome 53 doesn't necessarily fix it. Does anyone understand the release procedure of Chrome well enough to know if the linked patch above is already in Chrome 53 or not?

> Is this a precise statement of the workaround? "Any tag with a dynamic attribute must have at least one sibling."

I'm not sure it's that clear what the exact mechanism of the issue is, unfortunately. 

I suspect the issue is in the call to `HTML.isConstructedObject()` [here](https://github.com/meteor/meteor/blob/fix-pr-6889/packages/htmljs/html.js#L22) (you'll notice the [implementation of that function](https://github.com/meteor/meteor/blob/master/packages/htmljs/html.js#L188-L191) has some typeof checks that are *similar* to the chrome bug listed).

If we can't rely on that call then any nested tag is a potential source of problems---putting whitespace before every nested tag would be the only guaranteed workaround. However in the reproduction a loop was required to trigger the behavior, which makes sense given it's an optimization kicking in.

So I'd suggest a workaround would be to insert whitespace before nested tags inside loops in your templates.

However, if someone can reliably reproduce the problem using `%OptimizeFunctionOnNextCall()` (like [here](https://bugs.chromium.org/p/chromium/issues/detail?id=604033#c6)) then it may be possible to find a better working, perhaps by using a different implementation of `HTML.isConstructedObject`. I've been unable to achieve that though.

**Comment by nathan-muir** [comment]
Okay, did a bit more investigation.

1. [Get and compile v8 from source](https://github.com/v8/v8/wiki/Checking%20out%20source):
   ```sh
git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
PATH=`pwd`/depot_tools:"$PATH"
fetch v8
cd v8/
git checkout -b 5.1 -t branch-heads/5.1
gclient sync
make -j4 disassembler=on x64.release
   ```

2. [Create a minimal reproduction of the issue](https://gist.github.com/nathan-muir/7a530d488fddb1e6b335849d9f4284f8)

3. Verify the issue still exists

  ```sh
./out/x64.release/d8 repro.js --allow-natives-syntax --trace_hydrogen --trace_inlining
```

outputs something like

   ```
Concurrent recompilation has been disabled for tracing.
Did not inline HTMLTag called from make (target has context-allocated variables).
Did not inline HTMLTag called from make (target has context-allocated variables).
Inlined isConstructedObject called from HTMLTag.
Did not inline Array called from ArraySpeciesCreate (Dont inline [new] Array(n) where n isn't constant.).
broken, i=327
Inlined MaxSimple called from ArraySlice.
done!
```


----

**Findings**

It looks like this bug is separate (but similar) to the one linked up thread; that bug has been patched in the 5.1 branch, and should be released in Google Chrome.

I haven't been able to create a shorter reproduction, as the issue specifically arises when inlining `isConstructedObject` into `HTMLTag`.

If we force-optimise or never optimise `HTMLTag`, the test will pass.

**Next Steps**

If anyone can create a shorter reproduction, by forcing `isConstructedObject` to be inlined, then we might have a clear enough case for a bug report.

**Comment by nathan-muir** [comment]
Seems to be fine in the 5.2 branch

**Comment by mitar** [comment]
Fixes for this should probably go against https://github.com/meteor/blaze/tree/master/packages/htmljs and not code in this repository othewise we will fork the codebase.

**Comment by rijk** [comment]
In my case, I can confirm it's fixed in Chrome Canary (Version 54.0.2794.0 canary (64-bit)).

**Comment by rijk** [comment]
@tmeasday : with a breakpoint I've found the error to be triggered here: https://github.com/meteor/blaze/blob/f2e61cddd47fe3d7aab3a832d51ea9659138ec1f/packages/htmljs/visitors.js#L156

**Comment by rijk** [comment]
I am pretty sure this is caused by a bug in Chrome's optimiser. I've found it only happens when I start filtering while the page is still loading (so at that moment it's pretty heavy for the browser). Also, it does not happen anymore when I overwrite the `HTML.isConstructedObject` function from the console:

`var old = HTML.isConstructedObject; HTML.isConstructedObject = function (x) { return old( x ); }`

By the way, I've found filtering also going a LOT quicker after doing this, so there's definitely something going wrong big time when the optimiser kicks in.

**Comment by rijk** [comment]
I can also confirm this workaround, based on @nathan-muir's commit, fixed the issue:

```
window.HTML.isConstructedObject = function(x) { return $.isPlainObject(x); };
```

**Comment by Twisterking** [comment]
Unfortunately my knowledge of all this "behind the scenes stuff" is not good enough to help but I can show you my template which also causes the same issue for me.
It is a big table with an `#each` in it which fills the table with usually around **50-75 rows**. **Each row** looks like this:
```
<tr data-order_id="{{_id}}" class="{{#if canceled}}canceled{{/if}}">
	<td class="b"><a href="/view/{{_id}}" target="_blank"><span>{{getSupplierCompanyName supplier_id}}</span></a></td>
	<td><a href="/view/{{_id}}" target="_blank">{{name}}</a></td>			
	<td>{{formattime 'datetime' datetime}}</td>
	<td class="ar">{{itemcount}}</td>
	<td class="ac relative">
		<button class="standard btn_dropdown" data-toggle="dropdown">Mehr</button>
		<div class="dropdown">
			<ul>
				<li><a href="/view/{{_id}}" class="button clear btn_file_view" target="_blank">PDF anzeigen</a></li>
				<li><a href="/pdf/{{_id}}" class="button clear btn_file_download" target="_blank">PDF downloaden</a></li>
				{{#unless canceled}}
					<li><button class="clear text_red btn_cancelorder">Bestellung stornieren</button></li>
				{{else}}
					<li class="text red">Bestellung wurde storniert!</li>
				{{/unless}}
			</ul>
		</div>
	</td>
</tr>
```
As you can see "a lot" is happening in each row.
Usually, when I try to **access the url** which shows this template **directly**, the error **does occur**! Also another reload does not help. Which usually (not always) fixes it for me, is to load another url/view and then navigate to the other route via standard links/iron-router "redirects". Then it usually works!

I am on a Macbook, OSX 10.11.5 using Chrome 52.0.2743.75 beta (64-bit).

I really hope anyone on here can fix this! This bug is breaking part of my (production) app and makes my customers (and me) mad ... and I really don't know what to do to fix it ...

**Comment by brianlukoff** [comment]
Try putting a space between any two consecutive opening tags, like "<li> <a href...".

**Comment by Twisterking** [comment]
Thank you for this tip! I will try it and report back with news!

**Comment by rijk** [comment]
@brianlukoff appears to have fixed the problem in my case! 💯 👍

**Comment by brianlukoff** [comment]
Great!

**Comment by lowi** [comment]
Adding a white space before all <a> tags (adding &nbsp; was breaking the html layout) fix the issue with meteor 1.2 + blaze.

**Comment by Twisterking** [comment]
The white spaces did not fix it for me! :-( Can I provide you with any infos to help me? It happens again before trying to switch to a route which should show one of these "complicated" templates with the big `#each` in it. The URL in the address bar changes but the template is not shown and the OP console error appears. I really hope someone can help!

**Comment by tmeasday** [comment]
@Twisterking If I were you I'd try a "binary search" approach of deleting parts of your Blaze code above until you figure out which tag it is that's causing the problem. Chances are you just missed putting some whitespace in that complicated template.

It would be good to figure out a workaround that doesn't involve modifying templates.

Has anyone tried using @rijk's [monkey patch solution above](https://github.com/meteor/meteor/issues/6793#issuecomment-232939879) ?

- The original function that is getting mis-inlined is
```
window.HTML.isConstructedObject = function(t) {
    return t && "object" == typeof t && t.constructor !== Object && "function" == typeof t.constructor && t instanceof t.constructor
}
```

- @nathan-muir's version
```
window.HTML.isConstructedObject = function(x) { return !$.isPlainObject(x); };
```

- Although I wonder if that will give false positives for things that aren't objects at all. Maybe better is:

```
window.HTML.isConstructedObject = function(x) { return _.isObject(x) && !$.isPlainObject(x); };
```

**Comment by Twisterking** [comment]
@tmeasday : Hi there, long time no see haha. Thanks a lot for your reply! :-)
I am pretty sure that I didn't miss any whitespaces.
I also had a look into @nathan-muir's bugfix, but I was a bit confused on how to use it. on the client in the browser "window.HTML" is undefined and on the server "window" on it's own obviously is also undefined. So what I did is put the following into my **server side code**:
```
Meteor.startup(function () {
	// workaround for https://github.com/meteor/meteor/issues/6793 ??? maybe ...
	var old = HTML.isConstructedObject;
	HTML.isConstructedObject = function(x) { return old(x); }
});
```
Is there a better way to bugfix it maybe? Where/how do I have to use @nathan-muir's code?

**Comment by rijk** [comment]
@tmeasday I have to say, I reverted my monkey patch because it did give false positives, breaking other parts of the application. Your third version seems to have fixed that as well (luckily it's a bank holiday over here, so I was able to test it in production).

**Comment by Twisterking** [comment]
@rijk : How exactly did you apply the patch?

**Comment by rijk** [comment]
@Twisterking just put it somewhere in the client side code. I put in in my app container's `onRendered` function.

**Comment by Twisterking** [comment]
Does it really have to be in the `onRendered` callback? wouldn't `onCreated` be better?

**Comment by rijk** [comment]
Just try it out and see what works for you. I used `onRendered` to be sure the `isConstructedObject` was defined before I was overwriting it.

**Comment by tmeasday** [comment]
`HTML` is exported by the `htmljs` package which is implied by `blaze`. So you should see `window.HTML` if you are using blaze. I'm not sure why it would not be there. 

Alternatively, you might find it at `Package.htmljs.HTML`.

**Comment by alanning** [comment]
FYI, the "HTML.TransformingVisitor does not support foreign objects" error seems to still be happening with Chrome 52.  We see it pop up infrequently in production, 5 in the last week according to Kadira.  Chrome 51 and 52, Windows and OSX.  Updating all templates would be a challenge for our app; haven't tried the `isConstructedObject` monkey patch yet.

**Comment by nathan-muir** [comment]
@alanning I've had some success using this patch - https://github.com/nathan-muir/meteor/commit/b07ed32163e800bc9684a4a6c2fb4692da777542

As for Chrome 52; When I tested on v8 5.2 it was still in development, so it could've regressed.

Will do some more testing when available.

**Comment by Twisterking** [comment]
I am on Chrome 53 already and since I implemented the monkey patch, the error didn't occur anymore!

I implemented it like this:
```
Template.MasterLayout.onCreated(function () {
	window.HTML.isConstructedObject = function(x) { return _.isObject(x) && !$.isPlainObject(x); };
});
```
works for me!

**Comment by mitar** [comment]
I tried to [reproduce this](https://github.com/ecwyne/TransformingVisitor-Bug), but I could not on Mac OS X, 52.0.2743.116 (64-bit).

**Comment by ecwyne** [comment]
@mitar that is because i've added https://github.com/ecwyne/TransformingVisitor-Bug/commit/717cc7126064eddd3de95c53cddd06b9afd26f5f which provided a fix

I'll update the readme

**Comment by mitar** [comment]
I could not reproduce even earlier commit on Version 52.0.2743.116 (64-bit) or with Version 53.0.2785.113 (64-bit).

**Comment by dnprock** [comment]
I just saw this happen in our application, Chrome 53, meteor 1.4.1.1. No reproduce though.
![screen shot 2016-10-03 at 10 31 06 pm](https://cloud.githubusercontent.com/assets/497205/19064443/43c508e2-89c2-11e6-9799-0c2f8dba68c0.png)

**Comment by sferoze** [comment]
This happens rarely in my app in production. Meteor 1.4.1 Chrome 53. 

It happens when I switch routes. And the route doesn't load. 

But if I go to another route, and then back to the previous route, it loads then. 

Really strange, it happens once in awhile. 

When clicking a route, it just loads blank. And in the console its the same error.

**Comment by brianlukoff** [comment]
Try [this](https://github.com/meteor/meteor/issues/6793#issuecomment-233136203) (this fixed the issue completely for us).

**Comment by Twisterking** [comment]
Didn't help for me, I had to do this:
```
Template.MasterLayout.onCreated(function () {
    window.HTML.isConstructedObject = function(x) { return _.isObject(x) && !$.isPlainObject(x); };
});
```

**Comment by sferoze** [comment]
@Twisterking I have rarely seen this issue but it does come up once in awhile. I have no clue why. 

But I have added the code you posted

```
window.HTML.isConstructedObject = function(x) { return _.isObject(x) && !$.isPlainObject(x); };
```

On the client in a startup function and everything seems to still work in my app.

Just wondering if anyone has had any bad side effects due to this hack? 

Since it happens rarely for me, I am trying to judge if this hack is worth adding? It doesn't seem to be affecting anything negatively at the moment....

**Comment by elie222** [comment]
In my app on Chrome mobile Android, when I click a select box it no longer
opens on my phone. I wonder if it has anything to do with this.

**Comment by Twisterking** [comment]
@sferoze Nope no side effects on my side. have this in my production app for months now and the Bug never occurred again! :-)

**Comment by dnprock** [comment]
@brianlukoff thanks for the tip. We use jade so I couldn't add a space. I use <span> tag to separate <li> and <a> tag instead. Seems to work now.

**Comment by hwillson** [comment]
This issue has been moved to meteor/blaze#211 - we're reaching the issue comment limit in this repo thanks to this issue - kidding! 🙂 Seriously though - all Blaze issues are now being tracked via the [Blaze repo](https://github.com/meteor/blaze), so please continue the discussion there. Thanks everyone - closing here.
